# This script generates the ReportSummary and VOC Excel spreadsheets.
# After some post-generation manual formatting, the files are emailed to the DHHS and are uploaded to their SharePoint.
# The script requires as input:
# 1 - a sample list exported from Benchling as a CSV and renamed "Sample_List.csv"
# 2 - the text file generated by the wgs_script_v05.sh

library(reshape2)
library(writexl)

# -----------------------------------------------------------------
# define run-specific variables
# -----------------------------------------------------------------

SeqRun <- c("053")
# BatchDate <- c("24/4/23")
# use the following command to get today's date
BatchDate <- format(Sys.Date(), "%d/%m/%y")

# -----------------------------------------------------------------
# get data
# -----------------------------------------------------------------

# import Sample List file exported from Benchling
sample_info <- read.csv("Sample_List.csv")[, c(2,6,7,8,10:14)]

# get "rep" info
sample_info$rept <- substr(sample_info$WEHI_ID, 9,12)

sample_info$rept[sample_info$rept == ""] <- "0"
sample_info$rept[sample_info$rept == "rep1"] <- "1"
sample_info$rept[sample_info$rept == "rep2"] <- "2"
sample_info$rept[sample_info$rept == "rep3"] <- "3"
sample_info$rept[sample_info$rept == "rep4"] <- "4"
sample_info$rept[sample_info$rept == "rep5"] <- "5"
sample_info$rept[sample_info$rept == "NT"] <- "0"
sample_info$rept[sample_info$rept == "cat"] <- "1"

# shorten names
sample_info$Accession <- substr(sample_info$WEHI_ID, 1,8)

# calculate mean Cq
sample_info$Mean_Cq <- round(rowMeans(sample_info[,6:9]),1)

# import wgs summary file (rename as wgs_summary_a if merging two summaries together to make a final report)
wgs_summary_a <- read.table("wgs_run_summary.txt", sep = "\t", header = TRUE)

# import wgs summary file (concatenated samples)
wgs_summary_cat <- read.table("wgs_run_summary_53b.txt", sep = "\t", header = TRUE)

#merge wgs summary files
wgs_summary <- rbind(wgs_summary_a, wgs_summary_cat)

# shorten names
wgs_summary$Accession <- substr(wgs_summary$Accession, 1,8)

# remove samples from sample_info that were not sequenced 
sample_info <- subset(sample_info, sample_info$Accession %in% wgs_summary$Accession)

# set decimal places to 1
wgs_summary[,3:23] <- format(wgs_summary[,3:23], nsmall=1)

# ensure dataframes are in same order
wgs_summary <- wgs_summary[order(wgs_summary$Accession),]
sample_info <- sample_info[order(sample_info$Accession),]

# create column values
Purpose <- c("VOC - Routine")
SequencingQC <- c("Pass/Fail")
Recommendations <- NA
TechnicalNotes <- NA
Questionable <- NA
Repeat <- NA

# add columns
wgs_summary <- cbind(wgs_summary, Purpose, SeqRun, BatchDate, SequencingQC,Recommendations, TechnicalNotes, Questionable, Repeat)

# replace NAs with empty string
wgs_summary[is.na(wgs_summary)] <- ""

# add columns from sample info
wgs_summary <- cbind(wgs_summary, sample_info[,c(2:5,10,12)])

# set column names
colnames(wgs_summary) <- c("RNA_ID",
                           "Filtered Reads",
                           "AY %",
                           "BA1 %",
                           "BA2 %",
                           "BA3 %",
                           "BA4 %",
                           "BA5 %",
                           "BA275x %",
                           "BNx %",
                           "BQ1x %",
                           "BR2 %",
                           "CH11 %",
                           "XBB %",
                           "XBB15 %",
                           "XBB191 %",
                           "XBB116 %",
                           "XBC %",
                           "XBF %",
                           "Recombinant %",
                           "XBB192 %",
                           "XCC %",
                           "Unassigned %",
                           "Lineages Detected",
                           "Coverage Percent (%)",
                           "Depth",
                           "Purpose",
                           "Seq Run",
                           "Batch Date",
                           "Sequencing QC",
                           "Recommendations",
                           "Technical notes",
                           "Questionable",
                           "Repeat",
                           "DHHS SampleNo",
                           "Sample ID",
                           "DHHS SiteName",
                           "Retrieved",
                           "Repeat Number",
                           "Mean Cq")

# insert Technical notes
wgs_summary$`Technical notes`[wgs_summary$`Filtered Reads` < 275000 | wgs_summary$`Coverage Percent (%)` < 90 | wgs_summary$Depth < 1750] <- "QC criteria not met (Filtered Reads ≥ 275k; Coverage ≥ 90%; Depth ≥ 1750)"

# insert sequencing QC pass/fail comment
wgs_summary$`Sequencing QC`[wgs_summary$`Technical notes` == ""] <- "Pass"
wgs_summary$`Sequencing QC`[wgs_summary$`Sequencing QC` != "Pass"] <- "Fail"

# insert Recommendations
wgs_summary$Recommendations[wgs_summary$`Sequencing QC` == "Fail" & wgs_summary$`Mean Cq` > 37] <- "Sequencing failure - does not meet criteria for repeat testing (Cq < 37)"
wgs_summary$Recommendations[wgs_summary$`Sequencing QC` == "Fail" & wgs_summary$`Mean Cq` <= 37] <- "Sequencing failure - repeat sequencing"
wgs_summary$Recommendations[wgs_summary$`Sequencing QC` == "Fail" & wgs_summary$`Repeat Number` > 0] <- "Unable to sequence after multiple attempts"
wgs_summary$Recommendations[wgs_summary$`Sequencing QC` == "Pass" & wgs_summary$`Repeat Number` > 0] <- "Sample sequenced successfully upon repeat"
wgs_summary$Recommendations[wgs_summary$`Sequencing QC` == "Pass" & wgs_summary$`Mean Cq` > 37] <- "Cq is indicative of inhibition or low viral load which may adversely affect the quality and accuracy of the results"
wgs_summary$Recommendations[wgs_summary$`Sequencing QC` == "Pass" & wgs_summary$`Mean Cq` > 37 & wgs_summary$`Repeat Number` > 0] <- "Sample sequenced successfully upon repeat. Cq is indicative of inhibition or low viral load which may adversely affect the quality and accuracy of the results"

# insert repeat comment
wgs_summary$Repeat[wgs_summary$Recommendations == "Sequencing failure - repeat sequencing"] <- "Requested"

# insert Questionable comments
wgs_summary$Questionable[wgs_summary$`Sequencing QC` == "Fail" & wgs_summary$`Mean Cq` > 37] <- "Final - will not be repeated"
wgs_summary$Questionable[wgs_summary$`Sequencing QC` == "Fail" & wgs_summary$`Repeat Number` > 0] <- "Final - will not be repeated"
wgs_summary$Questionable[wgs_summary$Recommendations == "Sequencing failure - repeat sequencing"] <- "Pending - repeat requested"
wgs_summary$Questionable[wgs_summary$`Sequencing QC` == "Pass" & wgs_summary$`Repeat Number` > 0] <- "Final - after repeat"
# wgs_summary$Questionable[wgs_summary$`Sequencing QC` == "Pass" & wgs_summary$`Repeat Number` == 0] <- "Final"

# -----------------------------------------------------------------
# 'run report' file generation
# -----------------------------------------------------------------

# subset data for SummaryReport
wgs_run_report <- wgs_summary[, c("RNA_ID",
                                  "Repeat Number",
                                  "DHHS SampleNo",
                                  "Sample ID",
                                  "DHHS SiteName",
                                  "Retrieved",
                                  "Mean Cq",
                                  "Filtered Reads",
                                  "Coverage Percent (%)",
                                  "Depth",
                                  "Sequencing QC",
                                  "BA1 %",
                                  "BA2 %",
                                  "BA3 %",
                                  "BA4 %",
                                  "BA5 %",
                                  "BA275x %",
                                  "BNx %",
                                  "BQ1x %",
                                  "BR2 %",
                                  "CH11 %",
                                  "XBB %",
                                  "XBB15 %",
                                  "XBB191 %",
                                  "XBB116 %",
                                  "XBC %",
                                  "XBF %",
                                  "Recombinant %",
                                  "XBB192 %",
                                  "XCC %",
                                  "Unassigned %",
                                  "Lineages Detected",
                                  "Technical notes",
                                  "Repeat",
                                  "Recommendations",
                                  "Questionable")]


# export run report as Excel file
write_xlsx(wgs_run_report, "WGSseq053_20230731_SummaryReport.xlsx")

# -----------------------------------------------------------------
# 'power bi' file generation
# -----------------------------------------------------------------

# subset data for transformed VOC report
report_transform <- wgs_summary[, c("Purpose",
                                  "RNA_ID",
                                  "Seq Run",
                                  "DHHS SampleNo",
                                  "Batch Date",
                                  "Repeat Number",
                                  "BA1 %",
                                  "BA2 %",
                                  "BA3 %",
                                  "BA4 %",
                                  "BA5 %",
                                  "BA275x %",
                                  "BNx %",
                                  "BQ1x %",
                                  "BR2 %",
                                  "CH11 %",
                                  "XBB %",
                                  "XBB15 %",
                                  "XBB191 %",
                                  "XBB116 %",
                                  "XBC %",
                                  "XBF %",
                                  "Recombinant %",
                                  "XBB192 %",
                                  "XCC %",
                                  "Unassigned %",
                                  "Recommendations",
                                  "Technical notes",
                                  "Lineages Detected",
                                  "Filtered Reads",
                                  "Coverage Percent (%)",
                                  "Depth",
                                  "Questionable")]

# replace commas with pipes
report_transform$`Lineages Detected` <- gsub(x = report_transform$`Lineages Detected`, pattern = ",", replacement = " |")

# convert to long data format
report_transform <- melt(report_transform, id = (colnames(report_transform[grep(" %",colnames(report_transform),invert=T)])))

# remove " %" text
report_transform$variable <- gsub(x = report_transform$variable, pattern = " %",replacement = "")

# update column names
names(report_transform)[names(report_transform) == "variable"] <- "Variant"
names(report_transform)[names(report_transform) == "value"] <- "VariantPercent"

# subset & reorder
report_transform <- report_transform[, c("Purpose",
                                         "RNA_ID",
                                         "Seq Run",
                                         "DHHS SampleNo",
                                         "Batch Date",
                                         "Repeat Number",
                                         "Variant",
                                         "Recommendations",
                                         "Technical notes",
                                         "VariantPercent",
                                         "Lineages Detected",
                                         "Filtered Reads",
                                         "Coverage Percent (%)",
                                         "Depth",
                                         "Questionable")]

# set column names
colnames(report_transform) <- c("Purpose",
                                "RNA_ID",
                                "Seq Run",
                                "DHHS SampleNo",
                                "Batch Date",
                                "RepeatNumber",
                                "Variant",
                                "Recommendations",
                                "Technical notes",
                                "VariantPercent",
                                "Lineages Detected (%)",
                                "Filtered Reads",
                                "CoveragePercent",
                                "Depth",
                                "Questionable")

# export as excel file
write_xlsx(list(VOC_Results_v2 = report_transform), "20230731_053_VOC_ResultsWGS_cat.xlsx")
